### DISCUSSION TITLE
No stream pushed when using v4l2h264enc with libcamerasrc & gstreamer

### DISCUSSION THREAD

**Discussion by Topper69** [bluenviron/mediamtx#1475] [Questions]
Hi,

I want to use some gstreamer pipeline with libcamerasrc  & v4l2h264enc  (hardware accel H264 encoder) but it doesn't "register" and no stream if avaible (404).
My pipeline
<code>gst-launch-1.0 libcamerasrc ! video/x-raw,framerate=30/1,width=1280,height=720,format=I420 ! capssetter caps="video/x-h264, colorimetry=bt709" ! v4l2convert ! v4l2h264enc extra-controls="s,video_bitrate=600000,h264_i_frame_period=15,h264_profile=1,video_bitrate_mode=0" ! 'video/x-h264,level=(string)4' ! rtspclientsink location=rtsp://localhost:8554/test
</code>

Output of gstreamer:
<code>pi@baby:~ $ gst-launch-1.0 libcamerasrc ! video/x-raw,framerate=30/1,width=1280,height=720,format=I420 ! capssetter caps="video/x-h264, colorimetry=bt709" ! v4l2convert ! v4l2h264enc extra-controls="s,video_bitrate=600000,h264_i_frame_period=15,h264_profile=1,video_bitrate_mode=0" ! 'video/x-h264,level=(string)4' ! rtspclientsink location=rtsp://localhost:8554/test
Setting pipeline to PAUSED ...
[0:11:51.115818310] [2199]  INFO Camera camera_manager.cpp:299 libcamera v0.0.3+40-9b860a66
[0:11:51.162491605] [2200]  WARN RPI raspberrypi.cpp:1308 Mismatch between Unicam and CamHelper for embedded data usage!
[0:11:51.163742471] [2200]  INFO RPI raspberrypi.cpp:1425 Registered camera /base/soc/i2c0mux/i2c@1/imx219@10 to Unicam device /dev/media1 and ISP device /dev/media2
Pipeline is live and does not need PREROLL ...
Progress: (open) Opening Stream
Pipeline is PREROLLED ...
Prerolled, waiting for progress to finish...
Progress: (connect) Connecting to rtsp://localhost:8554/test
[0:11:51.174927556] [2204]  INFO Camera camera.cpp:1028 configuring streams: (0) 1280x720-YUV420
[0:11:51.175448146] [2200]  INFO RPI raspberrypi.cpp:805 Sensor: /base/soc/i2c0mux/i2c@1/imx219@10 - Selected sensor format: 1920x1080-SBGGR10_1X10 - Selected unicam format: 1920x1080-pBAA
Progress: (open) Retrieving server options
Progress: (open) Opened Stream
Setting pipeline to PLAYING ...
New clock: GstSystemClock
Progress: (request) Sending RECORD request
Redistribute latency...
</code>

On simple rtsp server log:
<code>2023/02/15 16:47:30 INF [RTSP] [conn [::1]:58534] opened
</code>

When i use a the videotestsrc, it's works:
<code>gst-launch-1.0 videotestsrc ! v4l2h264enc extra-controls="s,video_bitrate=600000,h264_i_frame_period=15,h264_profile=1" ! 'video/x-h264,level=(string)4' ! rtspclientsink location=rtsp://localhost:8554/test
</code>

When i pipe the raw/video stream from libcamera-vid on gstreamer fdsrc, i don't receive any stream on rtsp server.
<code>libcamera-vid --width 4096 --height 3040 -t 10000 --gain 4 --awbgains 1,1 --shutter 100000 --denoise cdn_fast --codec yuv420 -o - | \
gst-launch-1.0 fdsrc fd=0 ! rawvideoparse use-sink-caps=false width=4096 height=3040 format=i420 framerate=10/1 ! v4l2convert ! v4l2h264enc extra-controls="controls,video_bitrate_mode=0,video_bitrate=600000,h264_i_frame_period=15,repeat_sequence_header=1" ! 'video/x-h264,level=(string)4' ! rtspclientsink location=rtsp://localhost:8554/test
</code>

Any hit to make it works with libcamerasrc ?

**Comment by aler9** [comment]
Hello,
The server has no part in this and it's GStreamer that doesn't perform the RTSP handshake. From my experience, it happens when there's no data flowing through the pipeline.
My advice is to use the `rpiCamera` source feature of the server, that allows to publish the video of a Raspberry Camera without involving GStreamer or libcamera-vid.

[a comment by an automated account is not shown]
